[PATCH V7 0/20] Querying disk rotation speed and link type

cathay4t opened the pull request:
Hightlights:

 * ~~Query local disk rotation speed -- `lsm_scsi_rpm_of_disk_path()`.~~
 * Query local disk rotation speed -- `lsm_local_disk_rpm_get()`.
 * ~~Query local disk link type -- `lsm_scsi_link_type_of_disk_path()`.~~
 * Query local disk link type -- `lsm_local_disk_link_type_get()`.
 * ~~Use macro _wrapper() to remove duplicate code in `_scsi_clib.c`.~~ Already in master.
 * ~~New lsmcli command -- `lsmcli lsd` to list all local scsi disks.~~
 * New lsmcli command -- `lsmcli ldl` to list all local scsi disks.

FAQ:
 * Why not use udev for RPM which does not need root privilege?

   Udev does not support SAS disk RPM query yet.
   Should we create a patch to udev or use this SCSI standard way?

 * Why need root privilege?

   Required by SGIO IOCTL. 
   If kernel could expose more VPD pages (like 0xb1and 0x00) via sysfs like they do for VPD 0x83, 
   we don't need to do SGIO any more.

 * Why create your own SGIO function instead of using libsgutils2?

    IMHO, libsgutils2 is too heavy for our simple use here and SGIO is not that complex.

cathay4t commented:
@joehandzik Could you kindly check this pull request also? It seems you mentioned about retrieving
disk hardware information.

joehandzik commented:
Hey @cathay4t, just a quick message so you know I haven't forgotten about this...just swamped. I should be able to take a look tomorrow. Thanks for keeping me involved!

joehandzik commented:
Hey @cathay4t! So much for looking at this "tomorrow". :(

Anyway, I'm here now. First off, I think this is conceptually a really nice addition to lsm. You've effectively built a way for us to pull whatever VPD page we want from the disks, which is useful for several of my use cases. What do you still think you need to be comfortable enough to upstream this?

mulkieran commented:
Given that you always check for existence of a device when given a path it ```might``` be worthwhile to abstract that action into separate procedure.

The other issue is about running as root. Currently doc examples don't show, e.g., lsmcli being required to be run as root. Will that all have to change?

cathay4t commented:
@mulkieran Thanks for the suggestions. I will add document for root permission.
For lsmcli, I have uncommitted patch to add new command for local disk query.

joehandzik commented:
Hey @cathay4t, any timeline for this going upstream? This is probably half of the puzzle for what I'd still like to see in the libstoragemgmt infrastructure (the other half being SES functionality).

cathay4t commented:
@joehandzik I will try to include this in the 1.3 version which is plan to release on 1 May 2016.

joehandzik commented:
@cathay4t, ok, sounds fair. Any ideas on timeline for SES-like functionality? I have the functionality I need for one of the solutions we're working with in a fork, but a fork won't cut it for Ceph (IMO). I'm probably going to get my Ceph PR out in the open before the end of March, so I'd love to have an idea for how we would interact with SEPs and use SES in the upstream repo sooner rather than later. Do you have any ideas about how I could help you reach a design you're comfortable pushing upstream?

cathay4t commented:
@joehandzik I tried the SES ident and fault LED yesterday, I believe I could make it along with 1.3 (May 1 2016).

I created issue #91 to tracking this. Could take a look and let me know if that's not you are expecting.

cathay4t commented:
Patch set is ready for review.

tasleson commented:
@joehandzik Where is your forked version that contains the SES functionality you have for ceph?  I looked around and didn't see anything obvious, but maybe it's hiding in plain sight. @cathay4t I believe Joe's point is can we get some functionality in sooner, like the end of this month.

joehandzik commented:
@tasleson In your defense, I don't think I named it in an obvious way for folks outside of HPE...it's the 'wip-all' branch on my fork. https://github.com/HP-Scale-out-Storage/libstoragemgmt/tree/wip-all

My version has dependencies on lsscsi and sg_ses, neither of which I'm a huge fan of (I did it that way at first because I thought it would be adequate, and then we ended up actually using it for a Lustre plugin we've developed). The end result I have is more-or-less ideal to program around (and I have a working version using my libstoragemgmt fork in my Ceph fork), but I'm in favor of the way @cathay4t designed around VPD pages here. I think an ideal implementation for upstream consumption would have a more generic API around submitting SES commands (that would make it more powerful and extensible IMO). I haven't had the time to spend brainstorming a proposal for exactly what that would look like.

So while I do think we could have functionality in sooner than May if we just use what I have, I'm interested in seeing if we can do better than what I've already written. If doing better takes all the way to May but provides a really strong base of functionality that will work well for Ceph, Lustre, etc for years to come, I'll take the extra few month hit.

cathay4t commented:
@tasleson Thanks for the suggestions.

Both functions supports ATA disks as libata is wrapping ATA pages/words to SCSI pages, with rotation speed stored in SCSI VPD 0xB1 page.
Even more, the NVMe also provides SCSI VPD 0xB1 pages to indicate they are non-rotated disk.
I am indenting to rename all `lsm_scsi_XXX` functions to `lsm_block_XXX` to reflect that.

For SAN or HWRAID logical volume, the rotation speed has been set as unsupported which I guess is
provided by the array.

I am implementing the out of band query right now.

cathay4t commented:
I have moved some unrelated patches from this PR to #92 and will back to this PR after that.

Thank you.

cathay4t commented:
Rebased to current master and renamed functions to `lsm_local_disk_xxx`.

cathay4t commented:
V2 uploaded.
Changes:
 * Function renamed to `lsm_local_disk_xxx`.
 * Rebased to current master.

tasleson commented:
@cathay4t We need the ability to determine RPM for disks that are behind a RAID adapter and are not in JBOD mode.  I take it that will be in a different PR?

cathay4t commented:
Just noticed I could share some constants with plugin API.
Please let me put both library API and plugin API this single PR.

cathay4t commented:
V3 patch set uploaded.
Please merge PR #107 first.

Changes in V3:
 * Add plugin level API for disk rpm and link_type with support of MegaRAID and HPSA.
 * New constants `LSM_DISK_RPM_NO_SUPPORT` and `LSM_DISK_LINK_TYPE_NO_SUPPORT`.

cathay4t commented:
Rebased to current master.

cathay4t commented:
Add new patch to use pyudev instead of enumerate `/sys/block` folder.

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h:
```diff
@@ -73,6 +73,37 @@ int LSM_DLL_EXPORT lsm_local_disk_vpd83_get(const char *sd_path,
 
+/**
+ * Query the disk rotation speed - revolutions per minute(RPM) of given SCSI
+ * disk path.
```
Probably should remove `SCSI` as the device could be SATA too.  Should this be more generic term as storage doesn't have to be disk, NVMe etc.?

cathay4t replied:
Removed 'SCSI' and added '/dev/nvme0n1' in example.

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h:
```diff
@@ -73,6 +73,37 @@ int LSM_DLL_EXPORT lsm_local_disk_vpd83_get(const char *sd_path,
+/**
+ * Query the disk rotation speed - revolutions per minute(RPM) of given SCSI
+ * disk path.
+ * Require permission to open disk path(root user or disk group).
```
Requires

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h:
```diff
@@ -73,6 +73,37 @@ int LSM_DLL_EXPORT lsm_local_disk_vpd83_get(const char *sd_path,
+ * Require permission to open disk path(root user or disk group).
+ * New in version 1.3.
+ * @param[in]  disk_path
+ *                      String. The path of SCSI disk block, example "/dev/sdb".
```
Same as above

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h:
```diff
@@ -73,6 +73,37 @@ int LSM_DLL_EXPORT lsm_local_disk_vpd83_get(const char *sd_path,
+ *         LSM_ERR_LIB_BUG              When something unexpected happens.
+ *         LSM_ERR_NOT_FOUND_DISK       When provided disk path not found.
+ *         LSM_ERR_PERMISSION_DENIED    No sufficient permission to access
+ *                                      provided disk path.
```
remove `disk`

cathay4t replied:
Since we are `lsm_local_disk_xxx()`, I believe we could assume NVMe is also a disk. We could change to `lsm_local_block_xxx()` if that matters.

tasleson replied:
Lets leave it as disk.

tasleson commented on c_binding/lsm_local_disk.c:
```diff
@@ -751,3 +784,182 @@ int lsm_local_disk_vpd83_get(const char *disk_path, char **vpd83,
+/*
+ * Assuming vpd_0_len > 3 and vpd_0_data is not NULL, caller should make sure
+ * that.
+ */
```
caller should make sure of that.  

Or just re-word to: 

Preconditions: vpd_0_len > 3,  vpd_0_data is not NULL and vpd_0_len > ?

tasleson commented on c_binding/lsm_local_disk.c:
```diff
@@ -751,3 +784,182 @@ int lsm_local_disk_vpd83_get(const char *disk_path, char **vpd83,
+
+/*
+ * Assuming err_msg and vpd_data are not NULL, caller should make sure that.
+ */
```
Same as above

tasleson commented on c_binding/lsm_local_disk.c:
```diff
@@ -751,3 +784,182 @@ int lsm_local_disk_vpd83_get(const char *disk_path, char **vpd83,
+            _lsm_err_msg_set(err_msg,
+                             "BUG: VPD page 0x%02x is supported, "
+                             "but failed with error %d(%s)", ioctl_rc,
+                             strerror(ioctl_errno));
```
I'm thinking we should use the thread safe version of strerror.  The library isn't thread safe, but we should be trying to be thread aware.

tasleson commented on c_binding/lsm_local_disk.c:
```diff
@@ -751,3 +784,182 @@ int lsm_local_disk_vpd83_get(const char *disk_path, char **vpd83,
+    if (rc != LSM_ERR_OK)
+        goto out;
+
+    *rpm = (vpd_data[4] << 8) + vpd_data[5];
```
Wonder if it would be clearer to use: `ntohs`?

cathay4t replied:
`be32toh()` is better. Thanks for the suggestion.

cathay4t replied:
I improved these line by using C struct instead of hard coded magic number 4.

tasleson commented:
@cathay4t Sorry Gris, I didn't very far on my code review today, I will work on this first thing tomorrow morning.

cathay4t commented:
V4 uploaded.

Changes since V3:
 * Use C struct instead of magic number for rpm parsing from SCSI VPD page.
 * Use the thread-safe method `strerror_r()` to replace `strerror()`.

tasleson commented on c_binding/lsm_local_disk.c:
```diff
@@ -751,3 +804,214 @@ int lsm_local_disk_vpd83_get(const char *disk_path, char **vpd83,
+        goto out;
+    }
+
+    *rpm = be16toh(bdc->medium_rotation_rate_be);
```
I was thinking more like:

`be16toh(*((uint16_t *)&(vpd_data[4])));`

But having the structure works too.

tasleson commented on doc/man/lsmcli.1.in:
```diff
@@ -604,6 +604,11 @@ Required. Repeatable. Source file to clone (relative path).
 Required. Repeatable. Destination file for clone (relative path).
 
+.SS local-disk-list
+List all disks found on current local operation system.
```
operation -> operating

tasleson commented:
@cathay4t @joehandzik 

I have concerns with this commit: Add new command to list local disks. d87938a953f491a352040b34d27231e0828fe2c7

libStorageMgmt is a library that has a command line.  In this change we are adding functionality that cannot be utilized by a library user.  A user wanting to get this information would have to fork/exec the command line to get this information, something that is unacceptable IMHO.  We need to add this into the API and then expose it in the command line.

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h:
```diff
@@ -104,6 +104,56 @@ int LSM_DLL_EXPORT lsm_local_disk_vpd83_get(const char *sd_path,
+ *                                  Fibre Channel
+ *                              LSM_DISK_LINK_TYPE_SSA
+ *                                  Serial Storage Architecture, Old IBM tech.
+ *                              LSM_DISK_LINK_TYPE_SBP = 3,
```
The first enumerated types have no associated numerical value, but the rest do.  Please make them all the same, one way or the other.

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h:
```diff
@@ -104,6 +104,56 @@ int LSM_DLL_EXPORT lsm_local_disk_vpd83_get(const char *sd_path,
+ *                      be freed by lsm_error_free().
+ * @return LSM_ERR_OK                   On success.
+ *         LSM_ERR_INVALID_ARGUMENT     When any argument is NULL or
+ *                                      illegal sd_path or no enough permission.
```
or insufficient permission.

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_local_disk.h:
```diff
@@ -104,6 +104,56 @@ int LSM_DLL_EXPORT lsm_local_disk_vpd83_get(const char *sd_path,
+ *                                      illegal sd_path or no enough permission.
+ *         LSM_ERR_LIB_BUG              When something unexpected happens.
+ *         LSM_ERR_NOT_FOUND_DISK       When provided disk path not found.
+ *         LSM_ERR_PERMISSION_DENIED    No sufficient permission to access
```
Insufficient permission to access

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_types.h:
```diff
@@ -268,6 +268,21 @@ typedef enum {
+    LSM_DISK_LINK_TYPE_USB = 9,
+    LSM_DISK_LINK_TYPE_SOP = 0xa,
+    LSM_DISK_LINK_TYPE_PCIE = 0xb,
+} lsm_disk_link_type;
```
Any specific reason to make the last two hex, instead or remaining in base 10 like the others?

cathay4t replied:
Just use the value from T10 SPC pdf. I could change them to base 10 if that matters.

tasleson commented on c_binding/lsm_local_disk.c:
```diff
@@ -1015,3 +1023,89 @@ int lsm_local_disk_rpm_get(const char *sd_path, int32_t *rpm,
+
+/* Workflow:
+ *  * Query VPD supported pages, if ATA Information page is supported, the
+ *     we got a ATA.
```
Should this be, `then we have an ATA device`?

tasleson commented on test/tester.c:
```diff
@@ -3428,6 +3428,36 @@ START_TEST(test_local_disk_rpm_get)
+
+    rc = lsm_local_disk_link_type_get("/dev/sda", &link_type,  &lsm_err);
+    if (rc != LSM_ERR_OK)
+        lsm_error_free(lsm_err);
```
What is this call testing?  We aren't checking anything.

cathay4t replied:
Since we cannot tell whether '/dev/sda' exists or whether current user has sufficient permission, we
could never know what kind of error we could get.
But we should not get LIB_BUG for sure. Will update test case to check LIB_BUG.

tasleson commented:
@cathay4t The commit message for 35c3bf54d9e96925d3f0c1152f84a7238d83a6a6 for the plugin section, those should be `_set`, not `_get`

tasleson commented:
@cathay4t Overall this PR looks pretty good, mostly just a few small nits.  However, the local disk list in the command line, needs some discussion.

cathay4t commented:
V5 patch set uploaded.

Changes since V4:
 * New API `lsm_local_disk_list()` and `lsm.LocalDisk.list()` to get local disk paths.
 * Fixed issue mentioned by Tony above. Thank you Tony!
 * Fix `make check` on hpsa server with sda as RAIDed HDD logical volume(where rpm is not supported).

cathay4t commented:
@tasleson Could you review the V5 patch set? Thank you very much.

tasleson commented:
@cathay4t I looked over the list local disk API additions, looks good.  I'm not sure what else has changed.  I should have kept my local branch as is and compared to your new branch.  This is a big change set to fully review each time.  I'm going to do some testing to ensure it doesn't leak memory tomorrow, but I think it looks good.

tasleson commented:
@cathay4t Testing looks good, found no issues with memory on local disk list.

cathay4t commented:
@joehandzik Would you mind take a another look on this PR before we merge it?
Thank you.

cathay4t commented:
V6 patch set uploaded.

Changes in V6:
 * Add support of SAS SSD in HPSA plugin.

joehandzik commented on c_binding/lsm_local_disk.c:
```diff
@@ -751,3 +812,413 @@ int lsm_local_disk_vpd83_get(const char *disk_path, char **vpd83,
+}
+
+/* Workflow:
+ *  * Query VPD supported pages, if ATA Information page is supported, the
```
*then

joehandzik commented on doc/man/lsmcli.1.in:
```diff
@@ -654,6 +659,8 @@ Alias of 'access-group-add'
 .SS ad
 Alias of 'access-group-delete'
+.SS lsd
+Alias of 'local-scsi-disk-list'
```
Which is it, local-disk-list or local-scsi-disk-list?

joehandzik commented on test/cmdtest.py:
```diff
@@ -799,6 +799,13 @@ def volume_ident_led_clear_test(cap):
+    if os.geteuid() == 0:
+        call([cmd, 'local-disk-list'])
+    else:
+        print("Skipping test of 'local-scsi-disk-list' command when not "
```
Another leftover of scsi-disk-list, it would appear.

joehandzik commented on test/cmdtest.py:
```diff
@@ -799,6 +799,13 @@ def volume_ident_led_clear_test(cap):
+        call([cmd, 'local-disk-list'])
+    else:
+        print("Skipping test of 'local-scsi-disk-list' command when not "
+              "ran by root user")
```
*run

joehandzik commented:
@cathay4t I found a few minor spelling and naming mistakes, but once those are cleaned up you should be good to merge!

cathay4t commented:
V7 patch set uploaded.

Changes in V7:
 * Fix spelling and naming mistakes mentioned by Joe. Thanks Joe!
